vncviewer/GestureHandler.cxx

```diff
@@ -0,0 +1,364 @@
+
+    case 2:
+      // Not a single- or triple-touch gesture
+      this->state &= ~(GH_LEFTBTN | GH_MIDDLEBTN);
+      break;
+
+    case 3:
+      // Not a single- or double-touch gesture
+      this->state &= ~(GH_VSCROLL | GH_HSCROLL | GH_ZOOM);
+      this->state &= ~(GH_LEFTBTN | GH_RIGHTBTN);
+      break;
+
+    default:
+      this->state = GH_NOGESTURE;
+  }
+
+  if (hasDetectedGesture()) {
+    if ((tracked.size()) != 2 || (this->state != GH_RIGHTBTN))
+      pushEvent(GH_GestureBegin);
+  }
+}
+
+bool GestureHandler::hasDetectedGesture() {
+  // Invalid state if any of the undefined bits are set
+  if ((state & GH_UNDEFINED) != 0) {
+    return false;
+  }
+
+  // Check to see if the bitmask value is a power of 2
+  // (i.e. only one bit set). If it is, we have a state.
+  return state && !(state & (state - 1));
+}
+
+void GestureHandler::trackTouch(const XIDeviceEvent *ev) {
+  GHTouch ght;
+
+  ght.last_x = ght.first_x = ev->event_x;
+  ght.last_y = ght.first_y = ev->event_y;
+
+  tracked[ev->detail] = ght;
+
+  switch (tracked.size()) {
+    case 1:
+      break;
+
+    case 2:
+      this->state &= ~GH_LEFTBTN;
+      break;
+
+    case 3:
+      this->state &= ~(GH_RIGHTBTN | GH_VSCROLL | GH_HSCROLL | GH_ZOOM);
+      break;
+
+    default:
+      this->state = GH_NOGESTURE;
+  }
+
+  if (tracked.size() == 1)
+    timeoutTimer.start(GH_STTDELAY);
+}
+
+void GestureHandler::avgTrackedTouches(double *first_x, double *first_y,
+                                       double *last_x, double *last_y) {
+  size_t size = tracked.size();
+  double fx = 0, fy = 0, lx = 0, ly = 0;
+
+  std::map<int, GHTouch>::const_iterator iter;
+  for (iter = tracked.begin(); iter != tracked.end(); ++iter) {
+    fx += iter->second.first_x;
+    fy += iter->second.first_y;
+    lx += iter->second.last_x;
+    ly += iter->second.last_y;
+  }
+
+  if (first_x)
+    *first_x = fx / size;
+  if (first_y)
+    *first_y = fy / size;
+  if (last_x)
+    *last_x = lx / size;
+  if (last_y)
+    *last_y = ly / size;
+}
+
+bool GestureHandler::handleTimeout(rfb::Timer* t)
+{
+  if (t == &timeoutTimer)
+    touchTimeout();
+
+  return false;
+}
+
+void GestureHandler::updateTouch(const XIDeviceEvent *ev) {
+  GHTouch *touch;
+
+  // If this is an update for a touch we're not tracking, ignore it
+  if (tracked.count(ev->detail) == 0)
+    return;
+
+  touch = &tracked[ev->detail];
+
+  // Update the touches last position with the event coordinates
+  touch->last_x = ev->event_x;
+  touch->last_y = ev->event_y;
+
+  if (!hasDetectedGesture()) {
+    double h, v, fd, ld, dt;
+
+    // If the move is smaller than the minimum threshold, ignore it
+    if (std::abs(touch->first_x - ev->event_x) < GH_MTHRESHOLD &&
+        std::abs(touch->first_y - ev->event_y) < GH_MTHRESHOLD)
+      return;
+
+    // Because it's impossible to distinguish from a scroll, right
+    // click can never be initiated by a movement-based trigger.
+    this->state &= ~GH_RIGHTBTN;
```

Comment:
This is a good example of broken abstraction. This makes no sense unless you also remember that we're mapping two fingers to the right button. If the gesture had a name related to two fingers it would be obvious that is conflicts with other gestures using two fingers.

Reply:
Right. Let's rename this along with the rest of them as per your other comments.

Reply:
10d5a363